# 나스닥 기업 주가 예측
### 최소한 제출은 할 수 있도록 뼈대만 만들어 보았습니다. 

In [2]:
# 필요한 라이브러리를 아래 예시처럼 불러옵니다.
import pandas as pd
import numpy as np


In [4]:
# 해당 번호의 기업 주가 csv파일을 DataFrame 형태로 불러오는 함수입니다.
def load_file(org_idx):
    # return pd.read_csv(f"/mnt/elice/dataset/contest_data/org_{org_idx}.csv")
    return pd.read_csv(f"data/org_{org_idx}.csv")

# 15개의 기업 주가 파일을 DataFrame 형태로 하나씩 불러온 후 리스트에 저장하는 함수입니다.
def load_files():
    df_list = []
    for i in range(15):
        df_list.append(load_file(i+1))
    return df_list

org_3 = load_file(3)
print('3번 기업의 주가 데이터')
display(org_3.head())

3번 기업의 주가 데이터


Open      High       Low     Close  Adj Close    Volume
0  0.046412  0.044641  0.034246  0.044988   0.034071  0.209952
1  0.037014  0.048349  0.044059  0.049491   0.036170  0.187435
2  0.047310  0.051419  0.050161  0.041077   0.042702  0.158739
3  0.053429  0.046155  0.048430  0.042121   0.039320  0.138481
4  0.047967  0.053165  0.047416  0.047204   0.046467  0.117209

### 여기서 알고리즘을 만드시면 됩니다.
하나의 기업에 대한 주가 DataFrame을 입력값으로 받아, 10일치의 예측 결과를 리스트 형태로 출력하는 함수입니다.

In [3]:
# 한 기업의 주가 예측 결과 10일치를 리스트 형태로 반환하는 함수
def predict(df):
    # 간단한 예시를 위해 마지막 10일간의 주가를 거꾸로 뒤집은 결과를 얻는 함수를 만들었습니다.(주가는 생각했던 것과 거꾸로 가니까요...ㅠㅠ) 
    # 실제로는 머신러닝 및 딥러닝 알고리즘을 사용하여 값을 예측해야 합니다.
    predicted = df['Close'][-1:-11:-1] 
    return list(predicted)

print('3번 기업의 주가 예측 결과')
display(predict(org_3))

3번 기업의 주가 예측 결과


[0.4454438999931838,
 0.4501165102832787,
 0.4405543647275456,
 0.4495429847887275,
 0.4328229532296156,
 0.4307468663493707,
 0.4528269591033343,
 0.4581516279318614,
 0.4942340402942201,
 0.4830909104745027]

# 제출 및 채점

채점을 위해 결과 파일을 지시사항에 나온 형태로 만들어 `result.csv` 이름으로 저장합니다.

load_files() 함수로 데이터를 불러온 후 predict 함수를 활용하여 얻어낸 10일간의 예측값 리스트를 엮어서 csv 파일로 저장합니다. 

여러분이 만든 predict 함수의 출력 형식이 같다면 코드를 수정하지 않아도 됩니다.

In [4]:
# result.csv로 저장할 결과를 result_df라는 dataframe으로 만들어서 저장하는 예시
result_df = pd.DataFrame(
    columns=[f"org_{org_idx}" for org_idx in range(1, 16)], # org는 총 15개
    index=[f"day{day_idx}" for day_idx in range(1, 11)] # 예측 날짜는 총 10개
)

# 주가 csv 파일을 불러옵니다.
df_list = load_files()

# 각 기업별로 predict 함수를 적용하여 10일치의 예측 결과를 얻고, 이를 result_df에 저장
for i in range(15):
    result_df['org_{}'.format(i+1)] = predict(df_list[i])

# result_df를 results_csv로 저장
result_df.to_csv("result.csv")

display(result_df)

org_1     org_2     org_3     org_4     org_5     org_6     org_7  \
day1   0.629439  0.538555  0.445444  0.738810  0.776845  0.844508  0.752177   
day2   0.632841  0.524988  0.450117  0.740068  0.771690  0.833454  0.750236   
day3   0.628982  0.515033  0.440554  0.729721  0.783930  0.843107  0.732391   
day4   0.635360  0.525454  0.449543  0.725548  0.775677  0.832226  0.743948   
day5   0.589238  0.482530  0.432823  0.663816  0.714548  0.768056  0.701636   
day6   0.601934  0.495629  0.430747  0.692001  0.727308  0.806125  0.707838   
day7   0.632227  0.528284  0.452827  0.706959  0.745299  0.850886  0.712991   
day8   0.655650  0.538069  0.458152  0.716416  0.779343  0.869179  0.721019   
day9   0.671869  0.578098  0.494234  0.755932  0.774026  0.876039  0.744900   
day10  0.668488  0.556591  0.483091  0.741483  0.767223  0.858610  0.729666   

          org_8     org_9    org_10    org_11    org_12    org_13    org_14  \
day1   0.496297  0.149007  0.917413  0.835868  0.530321  0.466986  0.500812   
day2   0.493641  0.152784  0.903643  0.814073  0.535348  0.462789  0.502720   
day3   0.485511  0.161403  0.910775  0.824090  0.526767  0.477897  0.521021   
day4   0.473708  0.147293  0.889611  0.811480  0.524995  0.452009  0.493504   
day5   0.462637  0.141198  0.846167  0.741896  0.480084  0.429267  0.464961   
day6   0.447609  0.144039  0.879323  0.789337  0.492860  0.437755  0.497387   
day7   0.455080  0.144447  0.881253  0.811037  0.530535  0.461184  0.505268   
day8   0.446791  0.153708  0.906549  0.830661  0.537873  0.462882  0.489635   
day9   0.467488  0.178180  0.927330  0.851708  0.546720  0.509370  0.497680   
day10  0.474555  0.166155  0.925796  0.862000  0.534573  0.503021  0.517172   

         org_15  
day1   0.556041  
day2   0.571648  
day3   0.544717  
day4   0.513696  
day5   0.474561  
day6   0.486029  
day7   0.514641  
day8   0.528955  
day9   0.509843  
day10  0.507804

### 이제 채점을 해보겠습니다.
아래 코드를 수정하면 정상적인 채점을 수행할 수 없습니다.

In [5]:
import os

os.system('elice_grade result.csv')

Compressed 10240 bytes into 2762 bytes ==> 26.97%


send files ['result.csv'] for grade...
done!

Score: 0.043564
Duration: 1.0391626358032227 seconds
=== Message ===
[org_1] MAE = 0.036602
[org_2] MAE = 0.062465
[org_3] MAE = 0.029738
[org_4] MAE = 0.051958
[org_5] MAE = 0.038425
[org_6] MAE = 0.027560
[org_7] MAE = 0.026254
[org_8] MAE = 0.041807
[org_9] MAE = 0.023638
[org_10] MAE = 0.020147
[org_11] MAE = 0.026421
[org_12] MAE = 0.049354
[org_13] MAE = 0.041955
[org_14] MAE = 0.032378
[org_15] MAE = 0.144766

평균 MAE: 0.043564


0

### 마치며...

놀랍게도 최근 10일 값을 뒤집었을 뿐인데 평균 MAE 값이 0.043564로 나쁘지 않은 결과가 나왔습니다. (역시 주가는 생각과 거꾸로 가는걸까요 ㅠㅠ)

하지만 대회에서는 머신러닝과 딥러닝 알고리즘을 활용하여 예측값을 얻어내야 합니다.

이상으로 제출만을 위한 코드 소개를 마칩니다.

# 모두 화이팅!!